apps/web/ui/modals/add-edit-link-modal/og-section.tsx

```diff
@@ -28,8 +35,101 @@ export default function OGSection({
   setData: Dispatch<SetStateAction<LinkProps>>;
   generatingMetatags: boolean;
 }) {
+  const { id: workspaceId, exceededAI, mutate } = useWorkspace();
+
   const { title, description, image, proxy } = data;
 
+  const {
+    completion: completionTitle,
+    isLoading: generatingTitle,
+    complete: completeTitle,
+  } = useCompletion({
+    api: `/api/ai/completion?workspaceId=${workspaceId}`,
+    id: "metatags-title-ai",
+    onError: (error) => {
+      if (error.message.includes("Upgrade to Pro")) {
+        toast.custom(() => (
+          <UpgradeToProToast
+            title="You've exceeded your AI usage limit"
+            message={error.message}
+          />
+        ));
+      } else {
+        toast.error(error.message);
+      }
+    },
+    onFinish: (_, completion) => {
+      mutate();
+      va.track("Generated AI Meta Title", {
+        metadata: `Title: ${truncate(completion, 255 - (18 + data.url.length))} | URL: ${data.url}`,
+      });
+    },
+  });
+
+  const generateTitle = async () => {
+    completeTitle(
+      `You are an SEO expert. Generate an SEO-optimized meta title (max 120 characters) for the following URL:
+      
+      - URL: ${data.url}
+      - Meta title: ${data.title}
+      - Meta description: ${data.description}. 
+
+      Only respond with the title without quotation marks or special characters.
+      `,
+    );
+  };
+
+  useEffect(() => {
+    if (completionTitle) {
+      setData((prev) => ({ ...prev, title: completionTitle }));
+    }
+  }, [completionTitle]);
+
+  const {
+    completion: completionDescription,
+    isLoading: generatingDescription,
+    complete: completeDescription,
+  } = useCompletion({
+    api: `/api/ai/completion?workspaceId=${workspaceId}`,
+    id: "metatags-description-ai",
+    onError: (error) => {
+      if (error.message.includes("Upgrade to Pro")) {
+        toast.custom(() => (
+          <UpgradeToProToast
+            title="You've exceeded your AI usage limit"
+            message={error.message}
+          />
+        ));
+      } else {
+        toast.error(error.message);
+      }
+    },
```

Comment:
Feels like this logic could maybe be split out, or even included somehow in `useCompletion`

Reply:
agreed, this can definitely be optimized

Reply:
I also don't like that we're doing a `error.message.includes("Upgrade to Pro")` check, feels very fickle